common/client/node_lifecycle.go

```diff
@@ -322,46 +413,114 @@ func (n *node[CHAIN_ID, HEAD, RPC]) outOfSyncLoop(isOutOfSync func(num int64, td
+	var finalizedHeadsSub headSubscription[HEAD]
+	if n.chainCfg.FinalityTagEnabled() {
+		finalizedHeadsSub, err = n.registerNewSubscription(ctx, lggr.With("subscriptionType", "finalizedHeads"),
+			noNewFinalizedBlocksTimeoutThreshold, n.rpc.SubscribeToFinalizedHeads)
+		if err != nil {
+			lggr.Errorw("Subscribe to finalized heads failed on out-of-sync RPC node", "err", err)
+			n.declareUnreachable()
+			return
+		}
+
+		lggr.Tracew("Successfully subscribed to finalized heads feed on out-of-sync RPC node")
+		defer finalizedHeadsSub.Unsubscribe()
+	}
+
+	_, localHighestChainInfo := n.rpc.GetInterceptedChainInfo()
 	for {
+		if syncIssues == 0 {
+			// back in-sync! flip back into alive loop
+			lggr.Infow(fmt.Sprintf("%s: %s. Node was out-of-sync for %s", msgInSync, n.String(), time.Since(outOfSyncAt)))
+			n.declareInSync()
+			return
+		}
+
 		select {
 		case <-ctx.Done():
 			return
-		case head, open := <-ch:
+		case head, open := <-headsSub.Heads:
 			if !open {
-				lggr.Error("Subscription channel unexpectedly closed", "nodeState", n.getCachedState())
+				lggr.Errorw("Subscription channel unexpectedly closed", "nodeState", n.getCachedState())
 				n.declareUnreachable()
 				return
 			}
-			if !isOutOfSync(head.BlockNumber(), head.BlockDifficulty()) {
-				// back in-sync! flip back into alive loop
-				lggr.Infow(fmt.Sprintf("%s: %s. Node was out-of-sync for %s", msgInSync, n.String(), time.Since(outOfSyncAt)), "blockNumber", head.BlockNumber(), "blockDifficulty", head.BlockDifficulty(), "nodeState", n.getCachedState())
-				n.declareInSync()
-				return
+
+			if !n.onNewHead(lggr, &localHighestChainInfo, head) {
+				continue
+			}
+
+			syncIssues &= ^syncIssueHeadIsNotIncreasing
+			if outOfSync, _ := n.isOutOfSyncWithPool(localHighestChainInfo); !outOfSync {
+				syncIssues &= ^syncIssueNotInSyncWithPool
+			} else {
+				syncIssues |= syncIssueNotInSyncWithPool
+			}
+
+			if noNewHeadsTimeoutThreshold > 0 {
+				headsSub.ResetTimer(noNewHeadsTimeoutThreshold)
 			}
-			lggr.Debugw(msgReceivedBlock, "blockNumber", head.BlockNumber(), "blockDifficulty", head.BlockDifficulty(), "nodeState", n.getCachedState())
-		case <-time.After(zombieNodeCheckInterval(n.chainCfg.NodeNoNewHeadsThreshold())):
+
+			lggr.Debugw(msgReceivedBlock, "blockNumber", head.BlockNumber(), "blockDifficulty", head.BlockDifficulty(), "syncIssues", syncIssues)
+		case <-time.After(zombieNodeCheckInterval(noNewHeadsTimeoutThreshold)):
 			if n.poolInfoProvider != nil {
 				if l, _ := n.poolInfoProvider.LatestChainInfo(); l < 1 {
-					lggr.Critical("RPC endpoint is still out of sync, but there are no other available nodes. This RPC node will be forcibly moved back into the live pool in a degraded state")
+					lggr.Critical("RPC endpoint is still out of sync, but there are no other available nodes. This RPC node will be forcibly moved back into the live pool in a degraded state", "syncIssues", syncIssues)
```

Comment:
nit: think this should be Criticalw
```suggestion
					lggr.Criticalw("RPC endpoint is still out of sync, but there are no other available nodes. This RPC node will be forcibly moved back into the live pool in a degraded state", "syncIssues", syncIssues)
```